File: algoexpert/easy/2-valid-subsequence/2-valid-subsequence.js
// O(n) time | O(1) space
function isValidSubsequence(array, sequence) {
  let arrIdx = 0;
  let seqIdx = 0;
  while (arrIdx < array.length && seqIdx < sequence.length) {
    if (array[arrIdx] === sequence[seqIdx]) seqIdx++;
    arrIdx++;
  }
  return seqIdx === sequence.length;
}

console.log(
  isValidSubsequence([5, 1, 22, 25, 6, -1, 8, 10], [1, 6, -1, 10])
);

function isValidSubsequence2(array, sequence) {
  let seqIdx = 0;
  for (const value of array) {
    if (seqIdx === sequence.length) break;
    if (sequence[seqIdx] === value) seqIdx++;
  }
  return seqIdx === sequence.length;
}

console.log(
  isValidSubsequence2([5, 1, 22, 25, 6, -1, 8, 10], [1, 6, -1, 10])
);

File: algoexpert/easy/2-valid-subsequence/2-valid-subsequence.py
def isValidSubsequence(array, sequence):
    seqIdx = 0; 
    for value in array: 
        if seqIdx == len(sequence): 
            break
        elif value == sequence[seqIdx]:
            seqIdx += 1
    return seqIdx == len(sequence)        
        


print(isValidSubsequence([5, 1, 22, 25, 6, -1, 8, 10], [1, 6, -1, 10]))

def isValidSubsequence2(array, sequence):
    arrIdx = 0;
    seqIdx = 0;
    while arrIdx < len(array) and seqIdx < len(sequence):
        if array[arrIdx] == sequence[seqIdx]:
            seqIdx += 1
        arrIdx += 1
    
    return seqIdx == len(sequence)
    
        
print(isValidSubsequence2([5, 1, 22, 25, 6, -1, 8, 10], [1, 6, -1, 10]))